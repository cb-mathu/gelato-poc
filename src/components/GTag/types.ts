import { VueComponentProps } from '../../types';
import { HTMLAttributes, ExtractPropTypes } from 'vue';
import { TAG_PROPS } from './constants';

export enum TagVariant {
  Primary = 'primary',
  Secondary = 'secondary',
  Tertiary = 'tertiary',
}

export type TagProps = ExtractPropTypes<typeof TAG_PROPS>;
export type NativeSpanProps = Omit<HTMLAttributes, keyof TagProps>;
export type MergedTagProps = Partial<TagProps & VueComponentProps & NativeSpanProps>;
