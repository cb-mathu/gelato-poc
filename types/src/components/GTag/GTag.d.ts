import { MergedTagProps } from './types';
declare const _default: new () => {
    $props: MergedTagProps;
};
export default _default;
