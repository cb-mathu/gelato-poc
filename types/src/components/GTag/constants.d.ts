import { PropType } from 'vue';
import { TagVariant } from './types';
export declare const VARIANTS: TagVariant[];
export declare const TAG_PROPS: {
    variant: {
        type: PropType<TagVariant>;
        validator: (val: TagVariant) => boolean;
    };
    id: {
        type: StringConstructor;
        required: boolean;
    };
    class: {
        type: StringConstructor;
        required: boolean;
        default: string;
    };
};
