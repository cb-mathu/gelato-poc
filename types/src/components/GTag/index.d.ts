import GTag from './GTag';
import { TagVariant } from './types';
export { GTag, TagVariant };
