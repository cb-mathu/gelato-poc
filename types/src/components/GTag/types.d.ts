import { VueComponentProps } from '../../types';
import { HTMLAttributes, ExtractPropTypes } from 'vue';
import { TAG_PROPS } from './constants';
export declare enum TagVariant {
    Primary = "primary",
    Secondary = "secondary",
    Tertiary = "tertiary"
}
export declare type TagProps = ExtractPropTypes<typeof TAG_PROPS>;
export declare type NativeSpanProps = Omit<HTMLAttributes, keyof TagProps>;
export declare type MergedTagProps = Partial<TagProps & VueComponentProps & NativeSpanProps>;
