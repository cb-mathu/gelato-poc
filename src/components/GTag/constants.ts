import { DEFAULT_PROPS } from '../../utils/props';
import { PropType } from 'vue';
import { TagVariant } from './types';

export const VARIANTS = [TagVariant.Primary, TagVariant.Secondary, TagVariant.Tertiary];

export const TAG_PROPS = {
  ...DEFAULT_PROPS,
  variant: {
    type: String as PropType<TagVariant>,
    validator: (val: TagVariant) => VARIANTS.includes(val),
  },
};
