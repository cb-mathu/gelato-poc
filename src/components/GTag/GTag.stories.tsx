import GTag from './GTag';
import { MergedTagProps, TagVariant } from './types';
import { VARIANTS } from './constants';

export default {
  title: 'Component/GTag',
  component: GTag,
  argTypes: {
    variant: {
      description: 'Tag variants',
      options: VARIANTS,
      control: 'select',
    },
  },
};

const Template = (args: any) => ({
  setup() {
    return () => <GTag {...args}>Plan</GTag>;
  },
});

export const Story = Template.bind({});
export const DefaultTag = (props?: MergedTagProps) => <GTag {...(props as any)}>Plan</GTag>;
export const PrimaryTag = () => <GTag variant={TagVariant.Primary}>Plan</GTag>;
export const SecondaryTag = () => <GTag variant={TagVariant.Secondary}>Plan</GTag>;
export const TertiaryTag = () => <GTag variant={TagVariant.Tertiary}>Plan</GTag>;
