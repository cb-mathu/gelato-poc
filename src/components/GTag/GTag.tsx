import { defineComponent, SetupContext } from 'vue';
import { TAG_PROPS } from './constants';
import { MergedTagProps, TagProps, TagVariant } from './types';

const GTag = defineComponent({
  name: 'GTag',
  props: TAG_PROPS,

  setup(props: TagProps, { slots }: SetupContext) {
    return () => {
      const classList = {
        [`g-tag`]: true,
        [`g-tag--primary`]: props.variant === TagVariant.Primary,
        [`g-tag--secondary`]: props.variant === TagVariant.Secondary,
        [`g-tag--tertiary`]: props.variant === TagVariant.Tertiary,
        [`g-product-card__type`]: true,
      };

      return (
        <span {...props} class={classList}>
          {slots.default?.()}
        </span>
      );
    };
  },
});

export default GTag as {
  new (): {
    $props: MergedTagProps;
  };
};
