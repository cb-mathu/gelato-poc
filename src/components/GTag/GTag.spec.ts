import { render, screen, fireEvent } from '@testing-library/vue';
import * as stories from './GTag.stories';

const { DefaultTag, PrimaryTag, SecondaryTag, TertiaryTag } = stories;

describe('GTag : ', () => {
  test('renders primary tag with default args', () => {
    render(PrimaryTag());
    const tagElement = screen.getByText('Plan');
    expect(tagElement).toMatchSnapshot();
  });

  test('renders secondary tag with default args', () => {
    render(SecondaryTag());
    const tagElement = screen.getByText('Plan');
    expect(tagElement).toMatchSnapshot();
  });

  test('renders tertiary tag with default args', () => {
    render(TertiaryTag());
    const tagElement = screen.getByText('Plan');
    expect(tagElement).toMatchSnapshot();
  });

  test('renders default button with native attributes passed', () => {
    render(DefaultTag({ id: 'my-tag', class: 'my-tag', title: 'my tag' }));
    const tagElement = screen.getByText('Plan');
    expect(tagElement).toMatchSnapshot();
  });

  test('calls onClick prop when clicked', () => {
    const handleClick = jest.fn();
    render(DefaultTag({ onClick: handleClick }));
    fireEvent.click(screen.getByText('Plan'));
    expect(handleClick).toHaveBeenCalled();
  });
});
