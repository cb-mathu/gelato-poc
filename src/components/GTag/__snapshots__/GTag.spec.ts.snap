// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GTag :  renders default button with native attributes passed 1`] = `
<span
  class="my-tag g-tag g-product-card__type"
  id="my-tag"
  title="my tag"
>
  
  Plan
  
</span>
`;

exports[`GTag :  renders primary tag with default args 1`] = `
<span
  class="g-tag g-tag--primary g-product-card__type"
  variant="primary"
>
  
  Plan
  
</span>
`;

exports[`GTag :  renders secondary tag with default args 1`] = `
<span
  class="g-tag g-tag--secondary g-product-card__type"
  variant="secondary"
>
  
  Plan
  
</span>
`;

exports[`GTag :  renders tertiary tag with default args 1`] = `
<span
  class="g-tag g-tag--tertiary g-product-card__type"
  variant="tertiary"
>
  
  Plan
  
</span>
`;
